### ISSUE TITLE
Feature request: @return_future in ResultStorage put

### ISSUE THREAD

**Issue by RobertBiehl** [thumbor/thumbor#587]
Since thumbor 5 the result_storage get function is asynchroneous, I was wondering what kept the put function from receiving the same love.

https://github.com/thumbor/thumbor/blob/master/thumbor/result_storages/__init__.py#L42

I'd love if the put method would be changed from
````python
 class BaseStorage(object):
    ...
    def put(self, bytes):
        raise NotImplementedError()
````
to:
````python
class BaseStorage(object):
    ...
    @return_future
    def put(self, bytes, callback):
        raise NotImplementedError()

    @return_future
    def get(self, callback):
        raise NotImplementedError()
````

**Comment by SebastianBoldt** [comment]
+1

**Comment by norman-kapschefsky** [comment]
+1

**Comment by vvh-empora** [comment]
+1

**Comment by masom** [comment]
We updated Thumbor to fire the `put` in a IOLoop callback, preventing it from blocking requests.

If the backend supports async operations then this will be fully async.

See https://github.com/thumbor/thumbor/commit/b9f34378028382f1381745c93bd12b5d7c874e8d

**Comment by RobertBiehl** [comment]
Great I will look into it. Thank you!

**Comment by masom** [comment]
The metrics will currently be a little bit off so the method might need to yield with `maybe_future` and `return_future` as you suggested.

**Comment by heynemann** [comment]
Should we close this?

**Comment by RobertBiehl** [comment]
Ok, so as I understand there is no part in thumbor that waits for the confirmation of success/error of a put?

**Comment by masom** [comment]
Nope although right now if the `put` operation is async the recorded metrics will be wrong. I'll open a PR to call `yield maybe_future` on it.

**Comment by masom** [comment]
https://github.com/thumbor/thumbor/pull/590

**Comment by masom** [comment]
Closed by #590
